restapi/src/main/java/io/stargate/web/docsapi/service/ReactiveDocumentService.java

```diff
@@ -224,21 +228,32 @@ public ReactiveDocumentService(
                             false,
                             db.treatBooleansAsNumeric());
 
+                    Maybe<?> deleteBatch;
+
                     // dead leaf deletion init on non-empty collection
-                    if (!collector.isEmpty()) {
+                    if (deleteAuthorized && !collector.isEmpty()) {
                       logger.info(
                           String.format("Deleting %d dead leaves", collector.getLeaves().size()));
 
-                      // don't fail here if the unauthorized exception occurs, this is a read
-                      // request
-                      try {
-                        db.deleteDeadLeaves(
-                            namespace, collection, documentId, collector.getLeaves(), context, now);
-                      } catch (UnauthorizedException e) {
-                        logger.debug("Unauthorized when trying to delete dead leaves.", e);
-                      }
+                      deleteBatch =
```

Comment:
I would not wait for the deletion, the user issued a GET and we can simply do the deletion in the async way, non blocking on the response.. If it fails, the read would still be successful. We save both on time and on potential error:

```
Single.fromCallable(() -> db.authorizeDeleteDeadLeaves(namespace, collection))
   .filter(authenticated -> authenticated)
   .observeOn(Schedulers.io())
   .subscibeOn(Schedulers.io())
   .flatMap(_ -> RxUtils.singleFromFuture(() ->
                          db.deleteDeadLeaves(
                              namespace,
                              collection,
                              documentId,
                              now,
                              collector.getLeaves(),
                              context))
   .subscribe((r, t) -> {
    // log result or throwable
  })
   
```

Reply:
I provided the patch so we can quicker resolve this once you are online..


```
diff --git a/restapi/src/main/java/io/stargate/web/docsapi/service/ReactiveDocumentService.java b/restapi/src/main/java/io/stargate/web/docsapi/service/ReactiveDocumentService.java
index 2b0e44ff..2caaebea 100644
--- a/restapi/src/main/java/io/stargate/web/docsapi/service/ReactiveDocumentService.java
+++ b/restapi/src/main/java/io/stargate/web/docsapi/service/ReactiveDocumentService.java
@@ -26,6 +26,7 @@ import com.fasterxml.jackson.databind.node.ArrayNode;
 import com.fasterxml.jackson.databind.node.ObjectNode;
 import io.reactivex.rxjava3.core.Maybe;
 import io.reactivex.rxjava3.core.Single;
+import io.reactivex.rxjava3.schedulers.Schedulers;
 import io.stargate.auth.AuthenticationSubject;
 import io.stargate.auth.AuthorizationService;
 import io.stargate.auth.SourceAPI;
@@ -188,8 +189,6 @@ public class ReactiveDocumentService {
           authorizationService.authorizeDataRead(
               authenticationSubject, namespace, collection, SourceAPI.REST);
 
-          // Don't fail this read request if the corrective DELETE statements are not authorized,
-          // simply skip DELETE batches in that case (see the Rx pipeline below).
           boolean deleteAuthorized = db.authorizeDeleteDeadLeaves(namespace, collection);
 
           // we need to check that we have no globs and transform the stuff to support array
@@ -228,32 +227,44 @@ public class ReactiveDocumentService {
                             false,
                             db.treatBooleansAsNumeric());
 
-                    Maybe<?> deleteBatch;
-
                     // dead leaf deletion init on non-empty collection
-                    if (deleteAuthorized && !collector.isEmpty()) {
-                      logger.info(
-                          String.format("Deleting %d dead leaves", collector.getLeaves().size()));
-
-                      deleteBatch =
-                          Maybe.fromSingle(
-                              RxUtils.singleFromFuture(
-                                  () ->
-                                      db.deleteDeadLeaves(
-                                          namespace,
-                                          collection,
-                                          documentId,
-                                          now,
-                                          collector.getLeaves(),
-                                          context)));
-                    } else {
-                      deleteBatch = Maybe.just("delete not issued");
+                    if (!collector.isEmpty()) {
+                      // Don't fail this read request if the corrective DELETE statements are not
+                      // authorized,
+                      // simply skip DELETE batches in that case (see the Rx pipeline below).
+                      Single.fromCallable(() -> db.authorizeDeleteDeadLeaves(namespace, collection))
+                          .filter(
+                              auth -> {
+                                if (auth) {
+                                  logger.info(
+                                      String.format(
+                                          "Deleting %d dead leaves", collector.getLeaves().size()));
+                                } else {
+                                  logger.info(
+                                      String.format(
+                                          "No authority to delete %d dead leaves",
+                                          collector.getLeaves().size()));
+                                }
+                                return auth;
+                              })
+                          .observeOn(Schedulers.io())
+                          .subscribeOn(Schedulers.io())
+                          .flatMapSingle(
+                              any ->
+                                  RxUtils.singleFromFuture(
+                                      () ->
+                                          db.deleteDeadLeaves(
+                                              namespace,
+                                              collection,
+                                              documentId,
+                                              now,
+                                              collector.getLeaves(),
+                                              context)))
+                          .subscribe(
+                              r -> logger.info("Dead leaves successfully deleted."),
+                              t -> logger.warn("Error deleting dead leaves.", t));
                     }
 
-                    return deleteBatch.map(__ -> docsResult);
-                  })
-              .flatMap(
-                  docsResult -> {
                     // create json pattern expression if sub path is defined
                     if (!subDocumentPath.isEmpty()) {
                       String jsonPtrExpr =
```

Reply:
I did not want to change existing behaviour wrt surfacing DELETE errors, but you have a point. Since the DELETEs use explicit timestamps, and we only wait for the batch to be _accepted_ (it may not have been fully executed by the time we continue), running the DELETEs asynchronously is probably ok.

Reply:
Implemented that with some minor tweaks in 0bd2d386913abe87601f1174d7ac742a57d558cc